x-pack/plugins/integrations_manager/server/registry.ts

```diff
@@ -66,73 +19,35 @@ export async function fetchInfo(key: string) {
   return fetchJson(`${REGISTRY}/package/${key}`);
 }
 
-async function fetchUrl(url: string): Promise<string> {
-  return getResponse(url).then(responseToString);
-}
-
-async function fetchJson(url: string): Promise<object> {
-  const json = await fetchUrl(url);
-  const data = JSON.parse(json);
-  return data;
-}
-
-async function fetchArchiveBuffer(key: string): Promise<Buffer> {
-  return getResponse(`${REGISTRY}/package/${key}/get`).then(responseToBuffer);
-}
-
-async function filesFromBuffer(
-  buffer: Buffer,
-  predicate = (entry: Entry): boolean => true
-): Promise<string[]> {
-  const zipfile = await unzipFromBuffer(buffer);
-  const files: string[] = [];
-  zipfile.readEntry();
-  zipfile.on('entry', async (entry: Entry) => {
-    if (!predicate(entry)) return zipfile.readEntry();
-    if (cacheHas(entry.fileName)) return files.push(entry.fileName) && zipfile.readEntry();
-
-    const chunks: Buffer[] = [];
-    const stream: Readable = await new Promise((resolve, reject) =>
-      zipfile.openReadStream(entry, (err?: Error, readStream?: Readable) =>
-        err ? reject(err) : resolve(readStream)
-      )
-    );
-    stream.on('data', chunk => chunks.push(Buffer.from(chunk)));
-    stream.on('end', () => {
-      cacheSet(entry.fileName, Buffer.concat(chunks));
-      files.push(entry.fileName);
-      zipfile.readEntry();
-    });
-  });
-  return new Promise(resolve => zipfile.on('end', () => resolve(files)));
-}
-
-async function getFiles(
+export async function getArchiveInfo(
   key: string,
-  predicate = (entry: Entry): boolean => true
+  filter = (entry: ArchiveEntry): boolean => true
 ): Promise<string[]> {
   const archiveBuffer = await getOrFetchArchiveBuffer(key);
-  const files = await filesFromBuffer(archiveBuffer, predicate);
+  const extract = key.endsWith('.zip') ? unzipBuffer : untarBuffer;
+
+  const paths: string[] = [];
+  const onEntry = (entry: ArchiveEntry) => {
+    const { path, buffer } = entry;
+    paths.push(path);
+    if (cacheHas(path)) return;
+    if (buffer) cacheSet(path, buffer);
+  };
+
+  await extract(archiveBuffer, filter, onEntry);
 
-  return files;
+  return paths;
 }
 
 async function getOrFetchArchiveBuffer(key: string): Promise<Buffer> {
-  const archiveKey = `${key}-archive`;
-  if (!cacheHas(archiveKey)) {
-    await fetchArchiveBuffer(key).then(buffer => cacheSet(archiveKey, buffer));
+  if (!cacheHas(key)) {
+    await getResponseStream(`${REGISTRY}/package/${key}`)
+      .then(streamToBuffer)
+      .then(buffer => cacheSet(key, buffer));
```

Comment:
Just out of curiosity is there a reason to mix in the promise/then syntax with the await here?

Reply:
I don't think there's an issue having a `.then` to the right of `await`, but that bit could also be written as 

```typescript
async function getOrFetchArchiveBuffer(key: string): Promise<Buffer> {
  let buffer = cacheGet(key);
  if (!buffer) {
    buffer = await fetchArchiveBuffer(key);
    cacheSet(key, buffer);
  }

  if (buffer) {
    return buffer;
  } else {
    throw new Error(`no archive buffer for ${key}`);
  }
}

async function fetchArchiveBuffer(key: string): Promise<Buffer> {
  return getResponseStream(`${REGISTRY}/package/${key}`).then(streamToBuffer);
}
```

Reply:
I think that's more obvious. Happy to update it.